### ISSUE TITLE
Enhancement: [no-this-alias] should apply only to const, but not to let

### ISSUE THREAD

**Issue by EgonBolton** [typescript-eslint/typescript-eslint#7060]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-this-alias/

### Description

This rule makes sense for const, but not for let.
It is normal for a variable to take the initial value of this, and then use it to iterate.

### Fail

```typescript
class X {
  // ...
  methodX() {
    const current = this;
  }
}
```

### Pass

```typescript
class X {
  // ...
  methodX() {
    let current: this | null = this;
    while (current) {
      current = current.nextSibling;
      // ...
    }
  }
}
```

**Comment by bradzacher** [comment]
> It is normal for a variable to take the initial value of this, and then use it to iterate

Personally I wouldn't call that "normal" - it's a way to write code, for sure, but it's not a common practice. Self-referencing classes are far from a widely used pattern!

---

Not all codebases enforce `const` everywhere - so this change would make the rule useless in those codebases.

---

Additionally your hypothesis is that "all cases of assigning `this` to a `let` variable must be a correct iteration style" - which isn't the case. It's entirely possible to use this pattern in many other ways that aren't good practice.

Our recommendation is to use a disable comment for cases like this where you're intentionally assigning `this` to a variable to clearly annotate that you meant to alias it and not do something else.

For example in your code you might have intended to write `let current = this.nextSibling`, and starting at `this` was an error. A comment would signal clear intent.

**Comment by EgonBolton** [comment]
> For example in your code you might have intended to write let current = this.nextSibling, and starting at this was an error.

Well, I think you get the point.
```javascript
while (current) {
  // do something with current
  current = current.nextSibling;
}
```

> your hypothesis is that "all cases of assigning this to a let variable must be a correct iteration style" - which isn't the case.

No, my hypothesis is that dynamic variables are never aliases. An alias by definition is `const` IMO. 

> It's entirely possible to use this pattern in many other ways that aren't good practice.

I think that for the case where `let x = this` is never reassigned there is already the `prefer-const` rule. Can you think of an example where you need to initialize a dynamic variable `let x = this` and this rule makes sense? I can not 😅

Thank you very much for your time and your prompt response! ❤️
